src/util/snapshot.cpp

```diff
@@ -31,27 +37,128 @@ std::vector<SnapshotDiff> SnapshotData::getDirtyPages()
 std::vector<SnapshotDiff> SnapshotData::getChangeDiffs(const uint8_t* updated,
                                                        size_t updatedSize)
 {
-    // Work out which pages have changed in the comparison
+    // Work out which pages have changed
     size_t nThisPages = getRequiredHostPages(size);
     std::vector<int> dirtyPageNumbers =
       getDirtyPageNumbers(updated, nThisPages);
 
+    // Get iterator over merge regions
+    std::map<uint32_t, SnapshotMergeRegion>::iterator mergeIt =
+      mergeRegions.begin();
+
     // Get byte-wise diffs _within_ the dirty pages
+    //
     // NOTE - this will cause diffs to be split across pages if they hit a page
     // boundary, but we can be relatively confident that variables will be
     // page-aligned so this shouldn't be a problem
+    //
+    // For each byte we encounter have the following possible scenarios:
+    //
+    // 1. the byte is dirty, and is the start of a new diff
+    // 2. the byte is dirty, but the byte before was also dirty, so we
+    // are inside a diff
+    // 3. the byte is not dirty but the previous one was, so we've reached the
+    // end of a diff
+    // 4. the last byte of the page is dirty, so we've also come to the end of
+    // a diff
+    // 5. the byte is dirty, but is within a special merge region, in which
+    // case we need to add a diff for that whole region, then skip
+    // to the next byte after that region
     std::vector<SnapshotDiff> diffs;
     for (int i : dirtyPageNumbers) {
         int pageOffset = i * HOST_PAGE_SIZE;
 
-        // Iterate through each byte of the page
         bool diffInProgress = false;
         int diffStart = 0;
         int offset = pageOffset;
         for (int b = 0; b < HOST_PAGE_SIZE; b++) {
             offset = pageOffset + b;
             bool isDirtyByte = *(data + offset) != *(updated + offset);
-            if (isDirtyByte && !diffInProgress) {
+
+            bool isInMergeRegion =
+              mergeIt != mergeRegions.end() &&
+              offset >= mergeIt->second.offset &&
+              offset <= (mergeIt->second.offset + mergeIt->second.length);
```

Comment:
Minor point here, I think this should be a `<` not a `<=`.

If we have a merge region starting at offset 0 with length 3, we want to mark bytes 0,1,2 as dirty and belonging to this merge region. Byte 3 (i.e. 0 + 3) should be out of the region if I am not mistaken.

Reply:
Actually, after having gone through the code, afaict we will only have the case in which `offset == mergeIt->second.offset` as then we move the `offset` by `region.length`.

Maybe we could add an assertion after line 83 like:
```cpp
if (isDirtyByte && isInMergeRegion) {
    assert(offset == mergeIt->second.offset);
...
```